File: go/internal/tools/linode_profile_draft.go
package tools

import (
	"context"

	"github.com/mark3labs/mcp-go/mcp"

	"github.com/chadit/LinodeMCP/go/internal/config"
	linodev1 "github.com/chadit/LinodeMCP/go/internal/genpb/linode/mcp/v1"
	"github.com/chadit/LinodeMCP/go/internal/profiles"
	"github.com/chadit/LinodeMCP/go/internal/profiles/builder"
)

// draftProto converts a builder.Draft into its response message. The
// canonical serializer emits empty repeated fields as “[]“ not “null“,
// preserving the draft JSON contract of arrays over null.
func draftProto(draft *builder.Draft) *linodev1.ProfileDraftResponse {
	return &linodev1.ProfileDraftResponse{
		Name:                draft.Name,
		Description:         draft.Description,
		AllowedTools:        draft.AllowedTools,
		AllowedEnvironments: draft.AllowedEnvironments,
		RequiredTokenScopes: draft.RequiredTokenScopes,
		AllowYolo:           draft.AllowYolo,
	}
}

// ProfileDraftNewAnswer answers linode_profile_draft_new. It starts a new
// draft in the server's draft registry. Optional “clone_from“ seeds the draft
// from an existing profile (built-in or user-defined); without it the draft
// starts empty.
//
// Refusals, each answered as a tool result:
//
//   - msgDraftNameMissing when “name“ is empty
//   - "draft already exists" when the name is already drafted
//   - "clone_from profile not found" when “clone_from“ is non-empty
//     but resolves to no profile
//
// The clone source resolves against the config the server is running and the
// catalog the call's builder state carries, so a draft cloned mid-session sees
// the tools the running server has rather than the set it started with.
func ProfileDraftNewAnswer(
	ctx context.Context, request *mcp.CallToolRequest, cfg *config.Config,
) (*mcp.CallToolResult, error) {
	state, refusal := builderStateOrRefusal(ctx)
	if refusal != nil {
		return refusal, nil
	}

	name := request.GetString("name", "")
	if name == "" {
		return mcp.NewToolResultError(msgDraftNameMissing), nil
	}

	cloneFrom := request.GetString("clone_from", "")

	var source *profiles.Profile

	if cloneFrom != "" {
		resolved, ok := profiles.LookupProfile(cloneFrom, cfg, state.Catalog())
		if !ok {
			return mcp.NewToolResultError("clone_from profile not found: " + cloneFrom), nil
		}

		source = &resolved
	}

	// Create reports an empty name or a name the registry already holds, and
	// the empty name is refused above, so a failure here is the name already
	// being drafted. Reading the happy path first is what keeps that from
	// needing a branch no call can reach.
	draft, err := state.Drafts.Create(name, source)
	if err == nil {
		return MarshalProtoToolResponse(draftProto(draft))
	}

	return mcp.NewToolResultError("draft already exists: " + name), nil
}

// ProfileDraftShowAnswer answers linode_profile_draft_show. It reads the named
// draft and returns its current state. A miss refuses so the model can surface
// the typo or expired session.
func ProfileDraftShowAnswer(
	ctx context.Context, request *mcp.CallToolRequest, _ *config.Config,
) (*mcp.CallToolResult, error) {
	state, refusal := builderStateOrRefusal(ctx)
	if refusal != nil {
		return refusal, nil
	}

	name := request.GetString("name", "")
	if name == "" {
		return mcp.NewToolResultError(msgDraftNameMissing), nil
	}

	draft, ok := state.Drafts.Get(name)
	if !ok {
		return mcp.NewToolResultError(draftNotFound(name)), nil
	}

	return MarshalProtoToolResponse(draftProto(draft))
}

// ProfileDraftDiscardAnswer answers linode_profile_draft_discard. It removes
// the named draft from the registry. Idempotent: discarding a draft that is
// not there answers {"discarded": false} rather than a refusal, so the model
// can call it from cleanup paths without first checking existence.
func ProfileDraftDiscardAnswer(
	ctx context.Context, request *mcp.CallToolRequest, _ *config.Config,
) (*mcp.CallToolResult, error) {
	state, refusal := builderStateOrRefusal(ctx)
	if refusal != nil {
		return refusal, nil
	}

	name := request.GetString("name", "")
	if name == "" {
		return mcp.NewToolResultError(msgDraftNameMissing), nil
	}

	removed := state.Drafts.Discard(name)

	return MarshalProtoToolResponse(&linodev1.ProfileDraftDiscardResponse{
		Name:      name,
		Discarded: removed,
	})
}

File: python/src/linodemcp/tools/linode_profile_draft.py
"""Phase 8.3 draft lifecycle builder tools.

Three MCP tools wrap the in-memory draft registry from Phase 8.1:

- ``linode_profile_draft_new``: create a draft, optionally seeded from
  an existing profile via ``clone_from``.
- ``linode_profile_draft_show``: read a draft's current state.
- ``linode_profile_draft_discard``: remove a draft. Idempotent.

All three carry ``Capability.Meta`` so the profile filter always
admits them; they never touch the Linode API. The generator owns their
registration; what is left here is the body each one's answer hook calls. They
read the live registry off the builder state the server publishes for the call
(:mod:`linodemcp.tools.builderstate`); the clone source resolves against
the running config and that same state's catalog.
"""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, Any

from mcp.types import TextContent

from linodemcp.genpb.linode.mcp.v1 import profile_builder_pb2
from linodemcp.profiles.builder import (
    Draft,
    DraftExistsError,
)
from linodemcp.profiles.loader import lookup_profile
from linodemcp.tools.builderstate import (
    BUILDER_UNCONFIGURED,
    builder_state_from_context,
)
from linodemcp.tools.helpers import error_response
from linodemcp.tools.proto_response import serialize_api_response

if TYPE_CHECKING:
    from linodemcp.config import Config
    from linodemcp.profiles.profile import Profile


# The sentences the draft tools answer. Go spells them the same way, so a
# caller reads one refusal whichever binary served it.
NAME_MISSING = "name argument is required"


def draft_not_found(name: str) -> str:
    """The sentence every builder tool answers for a name no draft carries."""
    return f"draft not found: {name}"


# Argument-key constants. Used both in the schema and the handler so
# they can't drift.
_ARG_NAME = "name"
_ARG_CLONE_FROM = "clone_from"


def _draft_to_payload(draft: Draft) -> dict[str, Any]:
    """Serialize a Draft into the wire shape.

    The JSON tags match the Go side so cross-language tooling sees
    identical payloads. Empty lists serialize as ``[]`` not ``null``;
    the Draft dataclass already initializes lists, so no substitution
    is needed.
    """
    return {
        "name": draft.name,
        "description": draft.description,
        "allowed_tools": list(draft.allowed_tools),
        "allowed_environments": list(draft.allowed_environments),
        "required_token_scopes": list(draft.required_token_scopes),
        "allow_yolo": draft.allow_yolo,
    }


def profile_draft_new_result(
    arguments: dict[str, Any],
    cfg: Config,
) -> list[TextContent]:
    """Create a new draft and return its JSON representation.

    Refuses with a tool result, the way every other tool reports a bad
    argument, when the name is empty, the clone source resolves to nothing, or
    a draft already holds the name.
    """
    state = builder_state_from_context()
    if state is None:
        return error_response(BUILDER_UNCONFIGURED)

    name = arguments.get(_ARG_NAME, "")
    if not name:
        return error_response(NAME_MISSING)

    clone_from = arguments.get(_ARG_CLONE_FROM, "")

    source: Profile | None = None

    if clone_from:
        source = lookup_profile(clone_from, cfg, state.catalog())
        if source is None:
            return error_response(f"clone_from profile not found: {clone_from}")

    try:
        draft = state.drafts.create(name, source)
    except DraftExistsError:
        return error_response(f"draft already exists: {name}")

    result = serialize_api_response(
        _draft_to_payload(draft), profile_builder_pb2.ProfileDraftResponse()
    )
    return [TextContent(type="text", text=json.dumps(result, indent=2))]


def profile_draft_show_result(arguments: dict[str, Any]) -> list[TextContent]:
    """Read a draft and return its JSON representation.

    A missing name and a name no draft carries both refuse with a tool result
    the model can correct from.
    """
    state = builder_state_from_context()
    if state is None:
        return error_response(BUILDER_UNCONFIGURED)

    name = arguments.get(_ARG_NAME, "")
    if not name:
        return error_response(NAME_MISSING)

    draft = state.drafts.get(name)
    if draft is None:
        return error_response(draft_not_found(name))

    result = serialize_api_response(
        _draft_to_payload(draft), profile_builder_pb2.ProfileDraftResponse()
    )
    return [TextContent(type="text", text=json.dumps(result, indent=2))]


def profile_draft_discard_result(arguments: dict[str, Any]) -> list[TextContent]:
    """Remove a draft. Returns ``{name, discarded}``.

    ``discarded`` is True if the draft existed; False if not. Either case
    returns a normal response, so only an absent name refuses.
    """
    state = builder_state_from_context()
    if state is None:
        return error_response(BUILDER_UNCONFIGURED)

    name = arguments.get(_ARG_NAME, "")
    if not name:
        return error_response(NAME_MISSING)

    removed = state.drafts.discard(name)

    result = serialize_api_response(
        {"name": name, "discarded": removed},
        profile_builder_pb2.ProfileDraftDiscardResponse(),
    )
    return [TextContent(type="text", text=json.dumps(result, indent=2))]


__all__ = [
    "NAME_MISSING",
    "draft_not_found",
    "profile_draft_discard_result",
    "profile_draft_new_result",
    "profile_draft_show_result",
]
